hazelcast/src/main/java/com/hazelcast/jet/impl/connector/HazelcastReaders.java

```diff
@@ -336,19 +327,39 @@ public int getClassId() {
         }
     }
 
+    /**
+     * Create a ProcessorSupplier for a remote map in another cluster
+     */
     @Nonnull
-    public static <K, V, T> ProcessorSupplier readRemoteMapSupplier(
-            @Nonnull String mapName,
-            @Nonnull ClientConfig clientConfig,
-            @Nonnull Predicate<? super K, ? super V> predicate,
-            @Nonnull Projection<? super Entry<K, V>, ? extends T> projection
-    ) {
-        checkSerializable(Objects.requireNonNull(predicate), "predicate");
-        checkSerializable(Objects.requireNonNull(projection), "projection");
-
-        String clientXml = ImdgUtil.asXmlString(clientConfig);
-        return new RemoteProcessorSupplier<>(clientXml, new RemoteMapQueryReaderFunction<>(mapName, predicate,
-                projection));
+    public static <T, K, V> ProcessorSupplier readRemoteMapSupplier(RemoteMapSourceParams<T, K, V> params) {
+        if (params.hasDataSourceConnection()) {
+            if (params.hasPredicate()) {
+                checkSerializable(params.getPredicate(), "predicate");
+                checkSerializable(params.getProjection(), "projection");
+
+                RemoteMapQueryReaderFunction<K, V, T> readerSupplier = new RemoteMapQueryReaderFunction<>(
+                        params.getMapName(), params.getPredicate(), params.getProjection());
+
+                return RemoteProcessorSupplier.fromDataConnection(params.getDataConnectionName(), readerSupplier);
+            } else {
+                RemoteMapReaderFunction readerSupplier = new RemoteMapReaderFunction(params.getMapName());
+                return RemoteProcessorSupplier.fromDataConnection(params.getDataConnectionName(), readerSupplier);
+            }
+        } else {
+            String clientXml = ImdgUtil.asXmlString(params.getClientConfig());
+
+            if (params.hasPredicate()) {
```

Comment:
this branch is duplicated; I think it's safe to extract it a level above

Reply:
I agree it could be coded shorter but it is becoming less understandable. This way I it easier to see the flow